tests/gold_tests/null_transform/null_transform.test.py

```diff
@@ -0,0 +1,84 @@
+
+'''
+'''
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+
+import os
+Test.Summary = '''
+Test a basic null transform plugin
+'''
+
+# Need Curl
+Test.SkipUnless(
+    Condition.HasProgram("curl", "curl needs to be installed on system for this test to work")
+)
+Test.ContinueOnFail = True
+
+# Define default ATS
+ts = Test.MakeATSProcess("ts")
+server = Test.MakeOriginServer("server")
+
+Test.testName = ""
+request_header = {"headers": "GET / HTTP/1.1\r\nHost: www.example.com\r\n\r\n",
+                  "timestamp": "1469733493.993",
+                  "body": ""
+                  }
+# Expected response from origin server
+response_header = {"headers": "HTTP/1.1 200 OK\r\nConnection: close\r\n\r\n",
+                   "timestamp": "1469733493.993",
+
+                   "body": "This is expected response."}
+
+# Add response the server dictionary
+server.addResponse("sessionfile.log", request_header, response_header)
+ts.Disk.records_config.update({
+    'proxy.config.diags.debug.enabled': 1,
+    'proxy.config.diags.debug.tags': 'null_transform'
+})
+ts.Disk.remap_config.AddLine(
+    'map http://www.example.com http://127.0.0.1:{0}'.format(server.Variables.Port)
+)
+
+# Load plugin
+plugin_args = ""
+lib_dir = os.path.join(Test.Variables.AtsTestToolsDir, '../../lib')
+plugin_dir = ts.Env['PROXY_CONFIG_PLUGIN_PLUGIN_DIR']
+plugin_dir_src = os.path.join(Test.Variables.AtsTestToolsDir, 'plugins', 'null_transform.c')
+ts.Setup.Copy(plugin_dir_src, plugin_dir)
+
+in_basename = os.path.basename(plugin_dir_src)
+in_path = os.path.join(plugin_dir, in_basename)
+out_basename = os.path.splitext(in_basename)[0] + '.so'
+out_path = os.path.join(plugin_dir, out_basename)
+
+ts.Setup.RunCommand('tsxs -c {0} -o {1} -I {2}'.format(in_path, out_path, lib_dir))
+ts.Disk.plugin_config.AddLine("{0} {1}".format(out_basename, plugin_args))
+
+
+# www.example.com Host
+tr = Test.AddTestRun()
+tr.Processes.Default.Command = 'curl --proxy 127.0.0.1:{0} "http://www.example.com" --verbose -H "Proxy-Connection: keep-alive"'.format(
```

Comment:
I know we are relying on `curl`'s verbose output in many other tests, but I assert that relying on verbose output from `curl` is fundamentally broken as a concept. It is not a good or reliable way of serializing the data received from the traffic server, and is very error-prone. There is no expectation that this output will be consistent with differing versions of the curl tool, nor perhaps even with different invocations of the same version. Further, the gold file wildcard will match even new-line characters, and that makes it impossible to test certain cases (such as the absence of data). If we continue to use `curl`'s verbose output, I predict we will continue to have trouble with incorrect or fragile tests.

Another approach is to [read the bytes directly](https://github.com/apache/trafficserver/blob/7c5737b2e241045bbf16c01ba65968cfe273714e/tests/gold_tests/body_factory/http204_response.test.py#L81-L82) from the TCP connection to the traffic server—which is as consistent as the traffic server is—and do any necessary processing of that text.

Reply:
Well, I am not yet convinced we could write an expression to reliably predict all of the weird things `curl -v` might do in future versions.

Reply:
I think we can leave that for now since there are already a lot of tests  that use curl. I do agree with Derek that we need to look for some other way for verification like bytes comparison. I think traffic replay can be another option for verification which can verify the response field by field.